src/GPUActivityAgent.cpp

```diff
@@ -0,0 +1,460 @@
+
+                // Frequency selection is based upon the gpu compute activity.
+                // For active regions this means that we scale with the amount of work
+                // being done (such as SM_ACTIVE for NVIDIA GPUs).
+                //
+                // The compute activity is scaled by the GPU Utilization, to help
+                // address the issues that come from workloads have short phases that are
+                // frequency sensitive.  If a workload has a compute activity of 0.5, and
+                // is resident on the GPU for 50% of cycles (0.5) it is treated as having
+                // a 1.0 compute activity value
+                //
+                // For inactive regions the frequency selection is simply the efficient
+                // frequency from system characterization.
+                //
+                // This approach assumes the efficient frequency is suitable as both a
+                // baseline for active regions and and inactive regions. This is generally
+                // true of the efficient frequency is low power enough at idle due to clock
+                // gating or other hardware PM techniques.
+                //
+                // If f_efficient does not meet these criteria this behavior can still be
+                // achieved through tracking the GPU Utilization signal and setting frequency
+                // to a separate idle value (f_idle) during regions where GPU Utilizaiton is
+                // zero (or below some bar).
+                if (!std::isnan(gpu_utilization) &&
+                    gpu_utilization > 0) {
+                    gpu_utilization = std::min(gpu_utilization, 1.0);
+                    f_request = m_f_efficient + m_f_range * (gpu_compute_activity / gpu_utilization);
+                }
+                else {
+                    f_request = m_f_efficient + m_f_range * gpu_compute_activity;
+                }
+
+                // Tracking logic.  This is not needed for any performance reason,
+                // but does provide useful metrics for tracking agent behavior
+                if (gpu_compute_activity >= M_GPU_ACTIVITY_CUTOFF) {
+                    m_gpu_active_region_stop.at(domain_idx) = 0;
+                    if (m_gpu_active_region_start.at(domain_idx) == 0) {
+                        m_gpu_active_region_start.at(domain_idx) = m_time.value;
+                        m_gpu_active_energy_start.at(domain_idx) = m_gpu_energy.at(domain_idx).value;
+                    }
+                }
+                else {
+                    if (m_gpu_active_region_stop.at(domain_idx) == 0) {
+                        m_gpu_active_region_stop.at(domain_idx) = m_time.value;
+                        m_gpu_active_energy_stop.at(domain_idx) = m_gpu_energy.at(domain_idx).value;
+                    }
+                }
+            }
+
+            // Frequency clamping
+            f_request = std::min(f_request, m_f_max);
```

Comment:
Might want to add a `ifdef GEOPM_DBUG` with some prints that your frequency was clamped one way or the other unless it's easy to determine from the reports and traces.

Reply:
Added a "GPU Clipped Frequency Request" value to the report when GEOPM_DEBUG is defined.  I also moved the per GPU Region of interest start & stop time to the GEOPM_DEBUG case, as per GPU region time is enough for most non-debug cases.